src/main/java/org/apache/sling/api/resource/uri/ResourceUri.java

```diff
@@ -0,0 +1,184 @@
+
+import java.net.URI;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import org.apache.sling.api.request.RequestPathInfo;
+import org.apache.sling.api.resource.Resource;
+
+/**
+ * Represents an immutable URI that points to a resource or alternatively, can contain opaque URIs like {@code mailto:} or
+ * {@code javascript:}. Use {@link ResourceUri#adjust(Consumer)} or {@link ResourceUriBuilder} to create new or modified instances.
+ */
+public interface ResourceUri extends RequestPathInfo {
+
+    /**
+     * @return returns the URI.
+     */
+    public URI toUri();
+
+    /**
+     * @return returns the URI as String.
+     */
+    public String toString();
+
+    /**
+     * @return returns the scheme of the ResourceUri or null if not set
+     */
+    public String getScheme();
+
+    /**
+     * @return returns the user info of the ResourceUri or null if not set
+     */
+    public String getUserInfo();
+
+    /**
+     * @return returns the host of the ResourceUri or null if not set
+     */
+    public String getHost();
+
+    /**
+     * @return returns the port of the ResourceUri or null if not set
+     */
+    public int getPort();
+
+    /**
+     * @return returns the resource path or null if the URI does not contain a path.
+     */
+    @Override
+    public String getResourcePath();
+
+    /**
+     * @return returns the selector string or null if the URI does not contain selector(s)
+     */
+    @Override
+    public String getSelectorString();
+
+    /**
+     * @return returns the selector array (empty if the URI does not contain selector(s))
+     */
+    @Override
+    public String[] getSelectors();
+
+    /**
+     * @return returns the extension or null if the URI does not contain an extension
+     */
+    @Override
+    public String getExtension();
+
+    /**
+     * @return returns the path parameters or an empty Map if the URI does not contain any
+     */
+    public Map<String, String> getPathParameters();
+
+    /**
+     * @return returns the suffix or null if the URI does not contain a suffix
+     */
+    @Override
+    public String getSuffix();
+
+    /**
+     * @return returns the joint path of resource path, selectors, extension and suffix or null if resource path is not set
+     */
+    public String getPath();
+
+    /**
+     * @return returns the query part of the uri
+     */
+    public String getQuery();
+
+    /**
+     * @return returns the url fragment of the uri
+     */
+    public String getFragment();
+
+    /**
+     * The scheme-specific part of the URI, compare with Javadoc of class
+     * <h href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URI.html">URI</a>.
+     * 
+     * @return scheme specific part of the URI
+     */
+    public String getSchemeSpecificPart();
+
+    /**
+     * @return returns the corresponding suffix resource
+     */
+    @Override
+    public Resource getSuffixResource();
```

Comment:
As such an ResourceUri can also hold opaque URIs like ``mailto:foo@bar`` (according to the doc above), how does this manifest for this method?

Reply:
in the same way as if the suffix is not set (will return null). I will clarify in the Javadoc